### PR TITLE
Redirect to the identity registration with campaign code

### PR DESCRIPTION
**Pull request by chrisjowen**
Identity need to track better there referers so we now need to pass through a INTCMP query string parameter that will show where they came from and which tier they chose i.e. MEM_{PAGE}_{TIER}. 

See this card for all the permutations:
- https://trello.com/c/ZUTlcWIT/174-enable-identity-to-track-referrals-from-membership

### CODE REVIEW

File: frontend/app/actions/RegistrationUri.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    extractCampaignCode(request.headers(REFERRER_HEADER), request.path) match {
+      case Some(campaignCode) => redirectUrl.concat(s"&INTCMP=${campaignCode.get}")
+      case _ => redirectUrl
+    }
+  }
+
+  private def extractCampaignCode(referer: String, path: String) = {
+    val refererUrl = Uri.parse(referer)
+
+    val campaignReferer = refererUrl.path match {
```
Comment (joelochlann) [inline frontend/app/actions/RegistrationUri.scala]: Is it possible to get these strings programatically using Play's routing mechanism? That would make things easier if URLs change in future - hopefully these raw paths would only be stored in the `routes` file

File: frontend/test/actions/RegistrationUriTest.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      assertCodeFor("MEM_SUPUK_FRI", "/join/friend/enter-details", "https://membership.theguardian.com/supporter")
+    }
+
+  }
+
+
+  private def assertCodeFor(code: String, path: String, referrer: String) =  {
+    val request = new RequestBuilder().path(path).header("referer", referrer).build()._underlyingHeader()
+    val regUri: String = RegistrationUri.parse(request)
+    println(regUri)
```
Comment (joelochlann) [inline frontend/test/actions/RegistrationUriTest.scala]: Should this still be here? If it's to provide debug information when running the tests should it give a bit more info? e.g. `println(s"Registration URI: ${regUri}")` or something like that

**Comment by joelochlann** [comment]
:+1:

File: frontend/app/actions/RegistrationUri.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import configuration.Config
+import play.api.mvc.RequestHeader
+
+object RegistrationUri {
+  val REFERRER_HEADER: String = "referer"
+  val CAMPAIGN_SOURCE: String = "MEM"
+
+  def parse(request: RequestHeader) = {
+    val redirectUrl: String = Config.idWebAppRegisterUrl(request.uri)
+    extractCampaignCode(request.headers(REFERRER_HEADER), request.path) match {
```
Comment (rtyley) [inline frontend/app/actions/RegistrationUri.scala]: This has caused a lot of errors on the site, see:

https://app.getsentry.com/the-guardian/membership/group/90393824
https://app.getsentry.com/the-guardian/membership/group/90393829

File: frontend/app/actions/RegistrationUri.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      case Some(campaignCode) => redirectUrl.concat(s"&INTCMP=${campaignCode.get}")
+      case _ => redirectUrl
+    }
+  }
+
+  private def extractCampaignCode(referer: String, path: String) = {
```
Comment (rtyley) [inline frontend/app/actions/RegistrationUri.scala]: IntelliJ tells me that the return type of this method is `Option[Some[String]]` - that's indicative of poor Scala code.

In general, you don't want to nest `Option`s - you want to flatten them into a single `Option` with flatMap, or a for-comprehension. Also, `Some` being in the actual type signature indicates that it's redundant - the return type should just be `Option[String]`

**Pull request merged**